src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -990,247 +1067,423 @@ RouterSelectTask(Query *originalQuery, Query *query)
 	{
-		return true;
-	}
+		List *shardIntervalList = (List *) lfirst(prunedShardIntervalCell);
+		ShardInterval *shardInterval = NULL;
+		uint64 shardId = INVALID_SHARD_ID;
+		List *newPlacementList = NIL;
 
-	if (taskExecutorType != MULTI_EXECUTOR_REAL_TIME)
-	{
-		return false;
-	}
+		if (shardIntervalList == NIL)
+		{
+			continue;
+		}
 
-	Assert(commandType == CMD_SELECT);
+		Assert(list_length(shardIntervalList) == 1);
 
-	/*
-	 * Reject subqueries which are in SELECT or WHERE clause.
-	 * Queries which are recursive, with CommonTableExpr, with locking (hasForUpdate),
-	 * or with window functions are also rejected here.
-	 * Queries which have subqueries, or tablesamples in FROM clauses are rejected later
-	 * during RangeTblEntry checks.
-	 */
-	if (query->hasSubLinks == true || query->cteList != NIL || query->hasForUpdate ||
-		query->hasRecursive)
-	{
-		return false;
-	}
+		shardInterval = (ShardInterval *) linitial(shardIntervalList);
+		shardId = shardInterval->shardId;
 
-	if (query->groupingSets)
-	{
-		return false;
-	}
+		/* retrieve all active shard placements for this shard */
+		newPlacementList = FinalizedShardPlacementList(shardId);
 
-	/* only hash partitioned tables are supported */
-	distributedTableId = ExtractFirstDistributedTableId(query);
-	partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
-	partitionMethod = PartitionMethod(distributedTableId);
+		if (firstShard)
+		{
+			firstShard = false;
+			currentPlacementList = newPlacementList;
+		}
+		else
+		{
+			/* keep placements that still exists for this shard */
+			currentPlacementList = IntersectPlacementList(currentPlacementList,
+														  newPlacementList);
+		}
 
-	if (partitionMethod != DISTRIBUTE_BY_HASH)
-	{
-		return false;
+		/*
+		 * Bail out if placement list becomes empty. This means there is no worker
+		 * containing all shards referecend by the query, hence we can not forward
+		 * this query directly to any worker.
+		 */
+		if (currentPlacementList == NIL)
+		{
+			break;
+		}
 	}
 
-	/* extract range table entries */
-	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+	return currentPlacementList;
+}
 
-	/* query can have only one range table of type RTE_RELATION */
-	if (list_length(rangeTableList) != 1)
-	{
-		return false;
-	}
 
-	rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
-	if (rangeTableEntry->rtekind != RTE_RELATION)
+/*
+ * IntersectPlacementList performs placement pruning based on matching on
+ * nodeName:nodePort fields of shard placement data. We start pruning from all
+ * placements of the first relation's shard. Then for each relation's shard, we
+ * compute intersection of the new shards placement with existing placement list.
+ * This operation could have been done using other methods, but since we do not
+ * expect very high replication factor, iterating over a list and making string
+ * comparisons should be the fastest way around.
```

Comment:
Sorry, but I don't agree in the very least that this "should be the fastest way around".  I'd bet a lot that this is about a magnitude slower than the solution I proposed.

Maybe replace that with "... should be sufficient.".